src/webgpu/util/f32_interval.ts

```diff
@@ -0,0 +1,629 @@
+  assert(!Number.isNaN(x), `flush not defined for NaN`);
+  assert(!Number.isNaN(y), `flush not defined for NaN`);
+  const x_values = roundedF32Values(x);
+  const y_values = roundedF32Values(y);
+  const x_inputs = new Set<number>([...x_values, ...x_values.map(flushSubnormalNumber)]);
+  const y_inputs = new Set<number>([...y_values, ...y_values.map(flushSubnormalNumber)]);
+  const intervals = new Set<F32Interval>();
+  x_inputs.forEach(inner_x => {
+    y_inputs.forEach(inner_y => {
+      intervals.add(fn(inner_x, inner_y));
+    });
+  });
+  return F32Interval.span(...intervals);
+}
+
+/** Convert a triplet of points to an interval of valid rounded values, both flushed and not flushed */
+function roundAndFlushTernaryToInterval(
+  x: number,
+  y: number,
+  z: number,
+  fn: TernaryToInterval
+): F32Interval {
+  assert(!Number.isNaN(x), `flush not defined for NaN`);
+  assert(!Number.isNaN(y), `flush not defined for NaN`);
+  assert(!Number.isNaN(z), `flush not defined for NaN`);
+  const x_values = roundedF32Values(x);
+  const y_values = roundedF32Values(y);
+  const z_values = roundedF32Values(z);
+  const x_inputs = new Set<number>([...x_values, ...x_values.map(flushSubnormalNumber)]);
+  const y_inputs = new Set<number>([...y_values, ...y_values.map(flushSubnormalNumber)]);
+  const z_inputs = new Set<number>([...z_values, ...z_values.map(flushSubnormalNumber)]);
+  const intervals = new Set<F32Interval>();
+  // prettier-ignore
+  x_inputs.forEach(inner_x => {
+    y_inputs.forEach(inner_y => {
+      z_inputs.forEach(inner_z => {
+        intervals.add(fn(inner_x, inner_y, inner_z));
+      });
+    });
+  });
+
+  return F32Interval.span(...intervals);
+}
+
+/** Convert a point to an interval correctly rounded around the point */
+export function correctlyRoundedInterval(n: number): F32Interval {
+  return roundAndFlushPointToInterval(n, (n: number) => {
+    assert(!Number.isNaN(n), `absolute not defined for NaN`);
+    return toInterval(n);
+  });
+}
+
+/** Convert a point to an interval of absolute error around the point */
+export function absoluteErrorInterval(n: number, rng: number): F32Interval {
+  rng = Math.abs(rng);
+  const impl = (impl_n: number): F32Interval => {
+    assert(!Number.isNaN(impl_n), `absolute not defined for NaN`);
+    return new F32Interval(impl_n - rng, impl_n + rng);
+  };
+
+  return runPointImpl(toInterval(n), impl);
+}
+
+/** Convert a point to an interval of n ULP around the point */
+export function ulpInterval(n: number, numULP: number): F32Interval {
+  numULP = Math.abs(numULP);
+  const ulp_flush = oneULP(n, true);
+  const ulp_noflush = oneULP(n, false);
+  const ulp = Math.max(ulp_flush, ulp_noflush);
+  return new F32Interval(n - numULP * ulp, n + numULP * ulp);
+}
+
+/** For a point operation run its impl over the input that may be an interval */
+function runPointImpl(x: F32Interval, impl: PointToInterval): F32Interval {
```

Comment:
It's worth noting that using this function on an interval only makes sense when the impl function is monotonic over the input interval.

Reply:
Will do when productionizing